include/dice/template-library/polymorphic_allocator.hpp

```diff
@@ -197,6 +197,583 @@ namespace dice::template_library {
+			switch (discriminant_) {
+				case discriminant_type::First: {
+					return std::allocator_traits<Allocator1<T>>::allocate(alloc1_, n);
+				}
+				case discriminant_type::Second: {
+					return std::allocator_traits<Allocator2<T>>::allocate(alloc2_, n);
+				}
+				default: {
+					assert(false);
+					__builtin_unreachable();
+				}
+			}
+		}
+
+		constexpr void deallocate(pointer ptr, size_t n) noexcept(noexcept(std::allocator_traits<Allocator1<T>>::deallocate(std::declval<Allocator1<T> &>(), ptr, n))
+																   && noexcept(std::allocator_traits<Allocator2<T>>::deallocate(std::declval<Allocator2<T> &>(), ptr, n))) {
+			switch (discriminant_) {
+				case discriminant_type::First: {
+					return std::allocator_traits<Allocator1<T>>::deallocate(alloc1_, ptr, n);
+				}
+				case discriminant_type::Second: {
+					return std::allocator_traits<Allocator2<T>>::deallocate(alloc2_, ptr, n);
+				}
+				default: {
+					assert(false);
+					__builtin_unreachable();
+				}
+			}
+		}
+
+		constexpr bool operator==(polymorphic_allocator const &other) const noexcept {
+			if (discriminant_ != other.discriminant_) {
+				return false;
+			}
+
+			switch (discriminant_) {
+				case discriminant_type::First: {
+					return alloc1_ == other.alloc1_;
+				}
+				case discriminant_type::Second: {
+					return alloc2_ == other.alloc2_;
+				}
+				default: {
+					assert(false);
+					__builtin_unreachable();
+				}
+			}
+		}
+		constexpr bool operator!=(polymorphic_allocator const &other) const noexcept {
+			if (discriminant_ != other.discriminant_) {
+				return true;
+			}
+
+			switch (discriminant_) {
+				case discriminant_type::First: {
+					return alloc1_ != other.alloc1_;
+				}
+				case discriminant_type::Second: {
+					return alloc2_ != other.alloc2_;
+				}
+				default: {
+					assert(false);
+					__builtin_unreachable();
+				}
+			}
+		}
+
+		// TODO: for some reason variant requires less things to be noexcept https://en.cppreference.com/w/cpp/utility/variant/operator%3D
+		// TODO: how are they doing it?
+		// TODO: same for other operators
+		friend constexpr void swap(polymorphic_allocator &lhs, polymorphic_allocator &rhs) noexcept(std::is_nothrow_move_assignable_v<Allocator1<T>>
+																									 && std::is_nothrow_destructible_v<Allocator1<T>>
+																									 && std::is_nothrow_move_constructible_v<Allocator1<T>>
+																									 && std::is_nothrow_move_assignable_v<Allocator2<T>>
+																									 && std::is_nothrow_destructible_v<Allocator2<T>>
+																									 && std::is_nothrow_move_constructible_v<Allocator2<T>>) {
+
+			switch (lhs.discriminant_) {
+				case discriminant_type::First: {
+					Allocator1<T> tmp{std::move(lhs.alloc1_)};
+
+					switch (rhs.discriminant_) {
+						case discriminant_type::First: {
+							lhs.alloc1_ = std::move(rhs.alloc1_);
+							rhs.alloc1_ = std::move(tmp);
+							break;
```

Comment:
in this case you could simply call
```suggestion
							std::swap(lhs.alloc1_, rhs.alloc1_);
							
							break;
```

tmp can be moved into rhs.discriminant_ Second case. 

analoge for the Second case of the lhs.discriminant_ switch